src/main/java/net/minecraftforge/items/ItemStackHandler.java

```diff
@@ -0,0 +1,185 @@
+            {
+                this.stacks[slot] = reachedLimit ? ItemHandlerHelper.copyStackWithSize(stack, stack.stackSize - limit) : stack;
+            }
+            else
+            {
+                existing.stackSize += reachedLimit ? limit : stack.stackSize;
+            }
+            onContentsChanged(slot);
+        }
+
+        return reachedLimit ? ItemHandlerHelper.copyStackWithSize(stack, limit) : null;
+    }
+
+    public ItemStack extractItem(int slot, int amount, boolean simulate)
+    {
+        if (amount == 0)
+            return null;
+
+        validateSlotIndex(slot);
+
+        ItemStack existing = this.stacks[slot];
+
+        if (existing == null)
+            return null;
+
+        int toExtract = Math.min(amount, existing.getMaxStackSize());
+
+        if (existing.stackSize <= toExtract)
+        {
+            if (!simulate)
+            {
+                this.stacks[slot] = null;
+                onContentsChanged(slot);
+            }
+            return existing;
+        }
+        else
+        {
+            if (!simulate)
+            {
+                this.stacks[slot] = ItemHandlerHelper.copyStackWithSize(existing, existing.stackSize - toExtract);
+                onContentsChanged(slot);
+            }
+
+            return ItemHandlerHelper.copyStackWithSize(existing, toExtract);
+        }
+    }
+
+    protected int getStackLimit(int slot, ItemStack stack)
+    {
+        return stack.getMaxStackSize();
+    }
+
+    @Override
+    public NBTTagCompound serializeNBT()
+    {
+        NBTTagList nbtTagList = new NBTTagList();
+        for (int i = 0; i < stacks.length; i++)
+        {
+            if (stacks[i] != null)
+            {
+                NBTTagCompound itemTag = new NBTTagCompound();
+                itemTag.setInteger("Slot", i);
+                stacks[i].writeToNBT(itemTag);
+                nbtTagList.appendTag(itemTag);
+            }
+        }
+        NBTTagCompound nbt = new NBTTagCompound();
+        nbt.setTag("Items", nbtTagList);
+        nbt.setInteger("Size", stacks.length);
+        return nbt;
+    }
+
+    @Override
+    public void deserializeNBT(NBTTagCompound nbt)
+    {
+        setSize(nbt.hasKey("Size", Constants.NBT.TAG_INT) ? nbt.getInteger("Size") : stacks.length);
+        NBTTagList tagList = nbt.getTagList("Items", Constants.NBT.TAG_COMPOUND);
+        for (int i = 0; i < tagList.tagCount(); i++)
+        {
+            NBTTagCompound itemTags = tagList.getCompoundTagAt(i);
+            int slot = itemTags.getInteger("Slot");
+
+            if (slot >= 0 && slot < stacks.length)
+            {
+                stacks[slot] = ItemStack.loadItemStackFromNBT(itemTags);
+            }
+        }
+        onLoad();
+    }
+
+    protected void validateSlotIndex(int slot)
+    {
+        if (slot < 0 || slot >= stacks.length)
+            throw new RuntimeException("Slot " + slot + " not in valid range - [0," + stacks.length + ")");
```

Comment:
Maybe use a special subclass of RuntimeException here, so it can be catched specifically. 
Also the possibility of this Exception should be documented in the methods that may throw it.